### ISSUE TITLE
Warn on ambiguous redirect

### ISSUE THREAD

**Issue by chutz** [koalaman/shellcheck#1136]
Older versions of bash error out with an ambiguous redirect (ie parameters passed after redirect). Newer versions of bash apparently accept it, and just pass the args to the command.

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
cmd > output.log --long-argument1 --long-argmuent2 param1 param2
```
This also works in newer bash versions (!?!)
```sh
#!/bin/bash
> file.txt echo one two three four
```
The file will contain "one two three four"

Both of these are just confusing and should be warned about.

#### Here's what shellcheck currently says:
$ shellcheck myscript
No issues detected!

#### Here's what I wanted or expected to see:
warn about an ambiguous redirect

**Comment by nihilus** [comment]
Since the parameters are not in order you will get an error ofc.

**Comment by chutz** [comment]
Not with bash 4.4, it just works

**Comment by nihilus** [comment]
No it wont work since your example is wrong.

**Comment by nihilus** [comment]
Unless --long-argument2 takes two parameters.

**Comment by chutz** [comment]
It does work on bash-4.4, the shell will pass the parameters to the command, as one would not expect. This is why I am suggesting that shellcheck warn on this, it's just confusing.

param1 param2 may just be parameters, like `ls -l --human-readable file1 file2`

**Comment by nihilus** [comment]
Ah, beats me. Yeah that is indeed ambiguous.

**Comment by chutz** [comment]
For added ambiguous-ness
```sh
#!/bin/bash
> file.txt echo foo bar baz
```
Will result in `foo bar baz` being in file.txt

**Comment by koalaman** [comment]
This behavior is POSIX and supported by all bash versions. "Ambiguous redirect" is instead the error you get when you redirect to a word that becomes more than one due to globbing or wordsplitting: 

    true > *
    foo="bar baz";  true > $foo

ShellCheck already warns about quoting, but it doesn't warn against redirecting to globs, which it really should. 

I'm not so sure about a general warning about not having redirections last though. Things like `echo >> file "foo bar"` makes lines align nicely and `< file grep foo` mimics the order of `cat file | grep foo`

**Comment by Dmole** [comment]
I'm in favor of "redirections last" a large amount of shellcheck is already about making valid code legible / sane and redirects anywhere but last are bad IMO...I'm also in favor of conditions first (in crazy languages that support "x... if y")
